kofserver/scorer.py

```diff
@@ -0,0 +1,159 @@
+# Copyright (c) 2020 HITCON Agent Contributors
+# See CONTRIBUTORS file for the list of HITCON Agent Contributors
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# This file host the Scorer class, it is in charge of scoring the players.
+
+import logging
+import time
+
+from kofserver_pb2 import GameState
+from guest_agent_pb2 import ErrorCode as GuestErrorCode
+
+class Scorer:
+    def __init__(self, game):
+        # Initialize the variables from Game.
+        self.game = game
+        self.scenario = game.scenario
+        self.scoreboard = game.scoreboard
+        self.scanner = game.scanner
+        
+        # When's the last time we check the player's stats?
+        self.pidLastCheckTime = None
+        self.portLastCheckTime = None
+
+    # This is called by Game class when the game is started.
+    def NotifyGameStarted(self):
+        assert self.game.state == GameState.GAME_RUNNING
+        currentTime = time.time()
+        self.pidLastCheckTime = currentTime
+        self.portLastCheckTime = currentTime
+    
+    # Try to score the players. Return True is players are scored,
+    # False otherwise.
+    def TryScorePlayers(self):
+        if self.game.state != GameState.GAME_RUNNING:
+            logging.warn("Trying to score players when game is not running")
+            return False
+        
+        assert self.pidLastCheckTime is not None
+        assert self.portLastCheckTime is not None
+        
+        pidUptime = {}
+        portUptime = {}
+        playerDict = {} # Just a dictionary to store the players
```

Comment:
Note: Just Relace Dict with Set structure if the users are unique?

Reply:
Done